spec/memoize_query_spec.rb

```diff
@@ -612,4 +537,246 @@ def expect_to_eq_with_or_without_redis
+    # The following two relations are both SELECT * FROM Site where a <= 1 AND b = 2
+    let!(:relation3_with_comparator_and_other) { Site.where('a <= ?', 1).where(b: 2) }
+    let!(:relations_with_comparator_and_other) do
+      [
+        relation1_with_comparator_and_other, 
+        relation2_with_comparator_and_other, 
+        relation3_with_comparator_and_other,
+      ] 
+    end
+
+    # Deliberately use let below to let it be lazy-evaluated
+    # An update which will affect the results of the relation_with_comparator_and_other above
+    let(:update) { record.update(a: 2) }
+    # A creation which will not affect the results of the relation_with_comparator_and_other above
+    let(:unaffected_creation) { Site.create(b: 1) }
+
+    it_behaves_like 'only memoizes when there are bound conditions' do
+      let!(:result) { [record] }
+    end
+
+    it_behaves_like 'when update happens to the records' do 
+      let!(:result) { [] }
+    end
+
+    it_behaves_like 'with local cache' do 
+      let!(:unaffected_result) { [record] }
+      let!(:affected_result) { [] }
+    end
+  end
+
+  context 'when the queries have NOT' do 
+    let!(:record) { Site.create!(a: 1, b: 1) }
+    let!(:relation1_with_only_not) { Site.where.not(a: 2) }
+    let!(:relation2_with_only_not) { Site.where.not(a: 2, b: 2) }
+    let!(:relation3_with_only_not) { Site.where.not(a: 2).where.not(b: 1) }
+    let!(:relation_with_not_and_other) { Site.where.not(a: 2).where(b: 1) }
+    # where.not(a: ..., b:...) relation is treated differently in different Rails versions, 
+    # so we test its different behaviors in different versions next
+    # Details: https://bigbinary.com/blog/rails-6-deprecates-where-not-working-as-nor-and-will-change-to-nand-in-rails-6-1
+    let!(:special_relation_with_not) { Site.where.not(a: 2, b: 1) }
+
+
+    it 'does not memoize queries with only NOT' do
+      expect_not_to_use_redis do
+        expect(relation1_with_only_not.to_a).to eq([record])
+      end
+
+      expect_not_to_use_redis do
+        expect(relation2_with_only_not.to_a).to eq([record])
+      end
+
+      expect_not_to_use_redis do
+        expect(relation3_with_only_not.to_a).to eq([])
+      end
+    end
+
+    it 'does not memoize but adapts to different Rails versions flexibly' do
+      # We use ActiveRecord version to check Rails version
+      if ActiveRecord.version.to_s >= '6.1'
+        # Site.where.not(a: 2).where(b: 1) will be treated as NAND in Rails 6.1
+        expect(special_relation_with_not.to_a).to eq([record])
+      else
+        # Site.where.not(a: 2).where(b: 1) will be treated as NOR in and before Rails 6.0
+        # but will give a deprecation warning in Rails 6.0
+        expect(special_relation_with_not.to_a).to eq([])
+      end
+    end
+
+    it 'memoizes queries with both NOT and other bound conditions' do
+      expect_to_use_redis do
+        expect(relation_with_not_and_other.to_a).to eq([record])
```

Comment:
We should probably use `reload` all the time right? the test cases might not run in order.